From 97408c618724c5329599dbf5755f0bff3217219a Mon Sep 17 00:00:00 2001
From: Charlie Sharpsteen <source@sharpsteen.net>
Date: Mon, 17 Feb 2020 19:16:41 -0800
Subject: [PATCH 08/10] Add FFI libs to Debian 9 and 10 ARMHF

This commit adds libraries and build dependencies required to support
building the ffi rubygem on Debian armhf. These were added upstream in
puppet-agent release 6.13.0:

https://github.com/puppetlabs/puppet-runtime/commit/c8c87a6
---
 configs/components/_base-ruby-augeas.rb  |  8 ++---
 configs/components/_base-ruby-selinux.rb |  1 +
 configs/components/ruby-2.5.8.rb         |  3 ++
 configs/components/ruby-shadow.rb        |  2 ++
 configs/components/rubygem-ffi.rb        | 37 ++++++++++++++++++++++++
 configs/platforms/debian-10-armhf.rb     |  1 +
 configs/platforms/debian-9-armhf.rb      |  1 +
 7 files changed, 49 insertions(+), 4 deletions(-)

diff --git a/configs/components/_base-ruby-augeas.rb b/configs/components/_base-ruby-augeas.rb
index 5e7d6da..3834f4e 100644
--- a/configs/components/_base-ruby-augeas.rb
+++ b/configs/components/_base-ruby-augeas.rb
@@ -33,14 +33,14 @@ if platform.is_solaris?
     pkg.environment "RUBY", host_ruby
   end
   ruby = "#{host_ruby} -r#{settings[:datadir]}/doc/rbconfig-#{ruby_version}-orig.rb"
+elsif platform.is_cross_compiled_linux? && ( platform.name =~ /debian-(?:9|10)/ )
+  # Pick up SELinux configuration from ARM packages.
+  ruby = File.join(ruby_bindir, 'ruby')
+  pkg.environment "PKG_CONFIG_PATH", "#{File.join(settings[:libdir], 'pkgconfig')}:/usr/lib/#{settings[:platform_triple]}/pkgconfig"
 elsif platform.is_cross_compiled_linux?
   pkg.environment "RUBY", host_ruby
   ruby = "#{host_ruby} -r#{settings[:datadir]}/doc/rbconfig-#{ruby_version}-orig.rb"
   pkg.environment "LDFLAGS", settings[:ldflags]
-  if ( platform.name =~ /debian-(?:9|10)/ )
-    # Pick up SELinux configuration from ARM packages.
-    pkg.environment "PKG_CONFIG_PATH", "#{File.join(settings[:libdir], 'pkgconfig')}:/usr/lib/#{settings[:platform_triple]}/pkgconfig"
-  end
 else
   ruby = File.join(ruby_bindir, 'ruby')
 end
diff --git a/configs/components/_base-ruby-selinux.rb b/configs/components/_base-ruby-selinux.rb
index 58c8ad7..7af530b 100644
--- a/configs/components/_base-ruby-selinux.rb
+++ b/configs/components/_base-ruby-selinux.rb
@@ -58,6 +58,7 @@ end
 if platform.is_cross_compiled_linux? && ( platform.name =~ /debian-(?:9|10)/ )
   cc = "/usr/bin/#{settings[:platform_triple]}-gcc"
   system_include = "-I/usr/include"
+  ruby = "#{ruby_bindir}/ruby -rrbconfig"
 end
 
 pkg.build do
diff --git a/configs/components/ruby-2.5.8.rb b/configs/components/ruby-2.5.8.rb
index b20797e..5cf3845 100644
--- a/configs/components/ruby-2.5.8.rb
+++ b/configs/components/ruby-2.5.8.rb
@@ -192,6 +192,9 @@ component 'ruby-2.5.8' do |pkg, settings, platform|
   rbconfig_changes = {}
   if platform.is_aix?
     rbconfig_changes["CC"] = "gcc"
+  elsif platform.is_cross_compiled? && ( platform.name =~ /debian-(?:9|10)/ )
+    # Native Debian cross-compilation toolchain. No need to twiddle paths
+    # afterward.
   elsif platform.is_cross_compiled? || platform.is_solaris?
     rbconfig_changes["CC"] = "gcc"
     rbconfig_changes["warnflags"] = "-Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-constant-logical-operand -Wno-self-assign -Wunused-variable -Wimplicit-int -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wno-packed-bitfield-compat -Wsuggest-attribute=noreturn -Wsuggest-attribute=format -Wno-maybe-uninitialized"
diff --git a/configs/components/ruby-shadow.rb b/configs/components/ruby-shadow.rb
index 0490dbc..f9f13a2 100644
--- a/configs/components/ruby-shadow.rb
+++ b/configs/components/ruby-shadow.rb
@@ -11,6 +11,8 @@ component "ruby-shadow" do |pkg, settings, platform|
       pkg.environment "RUBY", settings[:host_ruby]
     end
     ruby = "#{settings[:host_ruby]} -r#{settings[:datadir]}/doc/rbconfig-#{settings[:ruby_version]}-orig.rb"
+  elsif platform.is_cross_compiled_linux? && ( platform.name =~ /debian-(?:9|10)/ )
+    ruby = File.join(settings[:ruby_bindir], 'ruby')
   elsif platform.is_cross_compiled_linux?
     pkg.environment "RUBY", settings[:host_ruby]
     ruby = "#{settings[:host_ruby]} -r#{settings[:datadir]}/doc/rbconfig-#{settings[:ruby_version]}-orig.rb"
diff --git a/configs/components/rubygem-ffi.rb b/configs/components/rubygem-ffi.rb
index 266cc53..846ac99 100644
--- a/configs/components/rubygem-ffi.rb
+++ b/configs/components/rubygem-ffi.rb
@@ -50,6 +50,43 @@ component "rubygem-ffi" do |pkg, settings, platform|
 
   if platform.name =~ /el-5-x86_64/
     pkg.install_file "/usr/lib64/libffi.so.5", "#{settings[:libdir]}/libffi.so.5"
+  elsif platform.is_cross_compiled? && ( platform.name =~ /debian-(?:9|10)/ )
+    pkg.environment "PKG_CONFIG_PATH", "#{File.join(settings[:libdir], 'pkgconfig')}:/usr/lib/#{settings[:platform_triple]}/pkgconfig"
+  elsif platform.is_cross_compiled?
+    ruby_api_version = settings[:ruby_version].gsub(/\.\d*$/, '.0')
+    base_ruby = case platform.name
+                when /solaris-10/
+                  "/opt/csw/lib/ruby/2.0.0"
+                else
+                  "/opt/pl-build-tools/lib/ruby/2.1.0"
+                end
+
+    # solaris 10 uses ruby 2.0 which doesn't install native extensions based on architecture
+    unless platform.name =~ /solaris-10/
+      # weird architecture naming conventions...
+      target_triple = if platform.architecture =~ /ppc64el|ppc64le/
+                        "powerpc64le-linux"
+                      elsif platform.name =~ /solaris-11-sparc/
+                        "sparc-solaris-2.11"
+                      else
+                        "#{platform.architecture}-linux"
+                      end
+
+      pkg.build do
+        [
+          %(#{platform[:sed]} -i 's/Gem::Platform.local.to_s/"#{target_triple}"/' #{base_ruby}/rubygems/basic_specification.rb),
+          %(#{platform[:sed]} -i 's/Gem.extension_api_version/"#{ruby_api_version}"/' #{base_ruby}/rubygems/basic_specification.rb)
+        ]
+      end
+    end
+
+    # make rubygems use our target rbconfig when installing gems
+    sed_command = %(s|Gem.ruby|&, '-r/opt/puppetlabs/puppet/share/doc/rbconfig-#{settings[:ruby_version]}-orig.rb'|)
+    pkg.build do
+      [
+        %(#{platform[:sed]} -i "#{sed_command}" #{base_ruby}/rubygems/ext/ext_conf_builder.rb)
+      ]
+    end
   end
 
   pkg.environment 'PATH', '/opt/freeware/bin:/opt/pl-build-tools/bin:$(PATH)' if platform.is_aix?
diff --git a/configs/platforms/debian-10-armhf.rb b/configs/platforms/debian-10-armhf.rb
index 87a277c..bbdab05 100644
--- a/configs/platforms/debian-10-armhf.rb
+++ b/configs/platforms/debian-10-armhf.rb
@@ -23,6 +23,7 @@ platform "debian-10-armhf" do |plat|
     'fakeroot',
     "libc6-dev:#{plat.get_architecture}",
     "libbz2-dev:#{plat.get_architecture}",
+    "libffi-dev:#{plat.get_architecture}",
     "libreadline-dev:#{plat.get_architecture}",
     "libselinux1-dev:#{plat.get_architecture}",
     'make',
diff --git a/configs/platforms/debian-9-armhf.rb b/configs/platforms/debian-9-armhf.rb
index dd45143..868e746 100644
--- a/configs/platforms/debian-9-armhf.rb
+++ b/configs/platforms/debian-9-armhf.rb
@@ -23,6 +23,7 @@ platform "debian-9-armhf" do |plat|
     'fakeroot',
     "libc6-dev:#{plat.get_architecture}",
     "libbz2-dev:#{plat.get_architecture}",
+    "libffi-dev:#{plat.get_architecture}",
     "libreadline-dev:#{plat.get_architecture}",
     "libselinux1-dev:#{plat.get_architecture}",
     'make',
-- 
2.27.0

